autopilot/contractor.go

```diff
@@ -58,34 +58,21 @@ func (c *contractor) currentPeriod() uint64 {
 	return c.currPeriod
 }
 
-func (c *contractor) contractSpending(contract worker.Contract) (bus.ContractSpending, error) {
-	// TODO: fetch contract hierarchy
-	var contracts []bus.Contract
-
-	// build a map
-	cmap := make(map[types.FileContractID]bus.Contract)
-	for _, contract := range contracts {
-		cmap[contract.ID] = contract
-	}
-
-	// fetch contract chain
-	curr, exists := cmap[contract.ID]
-	if !exists {
-		// TODO: return error here once fetching the history is implemented.
-		return bus.ContractSpending{}, nil // fmt.Errorf("contract with id '%v' not found", contract.ID)
+func (c *contractor) contractSpending(contract worker.Contract, currentPeriod uint64) (bus.ContractSpending, error) {
+	// Fetch past contracts in contract's history up until the start of the
+	// current period.
+	ancestors, err := c.ap.bus.AncestorContracts(contract.ID, currentPeriod)
+	if err != nil {
+		return bus.ContractSpending{}, err
 	}
-
 	// no history
-	if curr.RenewedFrom == (types.FileContractID{}) {
-		return curr.Spending, nil
+	if len(ancestors) == 0 {
+		return contract.Spending, nil
 	}
-
 	// compute total spending
-	total := curr.Spending
-	for exists {
-		if curr, exists = cmap[curr.RenewedFrom]; exists {
-			total = total.Add(curr.Spending)
-		}
+	total := contract.Spending
```

Comment:
I would extend the bus client with a helper that returns the total spending (and computes the total) alongside the `AncestorContracts` method